File: lc_plusplus/Reverse Vowels of a String.cc
// Write a function that takes a string as input and reverse only the vowels of a string.
//
// Example 1:
// Given s = "hello", return "holle".
//
// Example 2:
// Given s = "leetcode", return "leotcede".
//
// Note:
// The vowels does not include the letter "y".
//


class Solution {
public:
    string reverseVowels(string s) {
        string vowels{"aeiouAEIOU"};
        for (size_t i = 0, j = s.size(); i + 1 < j;) {
            if (vowels.find(s[i]) == string::npos) ++i;
            else if (vowels.find(s[j - 1]) == string::npos) --j;
            else swap(s[i++], s[j-- - 1]);
        }
        return s;
    }
};

File: lc_java/e/Reverse Vowels of a String.java
/**
 Write a function that takes a string as input and reverse only the vowels of a string.

 Example 1:
 Given s = "hello", return "holle".

 Example 2:
 Given s = "leetcode", return "leotcede".
**/

public class Solution {
	public String reverseVowels(String s) {
		String vowels = "aeiouAEIOU";
		char[] ca = s.toCharArray();
		int i = 0;
		int j = ca.length - 1;
		while (i < j) {
			if (vowels.indexOf(ca[i]) == -1) i++;
			else if (vowels.indexOf(ca[j]) == -1) j--;
			else {
				char tmp = ca[i];
				ca[i] = ca[j];
				ca[j] = tmp;
				i++;
				j--;
			}
		}
		return String.valueOf(ca);
	}
}

